### ISSUE TITLE
`Expected 0 positional arguments (reportCallIssue)` with explicitly typed `Callable[[int], SupportsInt]`

### ISSUE THREAD

**Issue by posita** [microsoft/pyright#11343]
**Describe the bug**
Attempting to use `f: Callable[[int], SupportsInt]` as `f(0)` after assignment from a `type(<SupportsInt>)` results in `Expected 0 positional arguments (reportCallIssue)` in apparent contradiction to the explicit argument in type declaration.

**Code or Screenshots**

```python
# pyright_test.py
from collections.abc import Callable
from typing import SupportsInt


def _test(init_val: SupportsInt, /) -> SupportsInt:
    mk: Callable[[int], SupportsInt] = type(init_val)
    return mk(int(init_val))  # <-- HERE
```

``` sh
% pyright --version
pyright 1.1.408
% pyright pyright_test.py
/.../pyright_test.py
  /.../pyright_test.py:8:15 - error: Expected 0 positional arguments (reportCallIssue)
1 error, 0 warnings, 0 informations
```

**Comment by dchevell** [comment]
FWIW I don't think this is a bug, it looks like expected behaviour.

`SupportsInt` is a protocol with only one method defined: `__int__`. It defines no `__init__`, therefore it inherits the default `object.__init__(self)` which takes no parameters. When you call `type(init_val)`, the type checker can't make any assumptions here - `init_val` could very well have a type that doesn't take any parameters. Adding `Callable[[int], SupportsInt]` doesn't add new type information that was missing from the original parameter type declaration, so it's not going to do what you're hoping for here. (Maybe that annotation should raise an error?)

I'd suggest solving your problem by extending the `SupportsInt` protocol to one that explicitly defines an init method:

```python
class CallableSupportsInt(SupportsInt):
    def __init__(self, __x: int) -> None: ...

def _test(init_val: CallableSupportsInt, /) -> SupportsInt:
    mk = type(init_val)
    return mk(int(init_val))
```

**Comment by posita** [comment]
Ah, interesting. This wasn't picked up by mypy, pyrefly, or ty, so I thought pyright may have been confused, but your explanation makes things clear(er). Thanks!
